plugins/meta/bandwidth/main.go

```diff
@@ -0,0 +1,199 @@
+
+	if err := json.Unmarshal(stdin, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse network configuration: %v", err)
+	}
+
+	// Parse previous result. Remove this if your plugin is not chained.
+	if conf.RawPrevResult != nil {
+		resultBytes, err := json.Marshal(conf.RawPrevResult)
+		if err != nil {
+			return nil, fmt.Errorf("could not serialize prevResult: %v", err)
+		}
+		res, err := version.NewResult(conf.CNIVersion, resultBytes)
+		if err != nil {
+			return nil, fmt.Errorf("could not parse prevResult: %v", err)
+		}
+		conf.RawPrevResult = nil
+		conf.PrevResult, err = current.NewResultFromResult(res)
+		if err != nil {
+			return nil, fmt.Errorf("could not convert result to current version: %v", err)
+		}
+	}
+	// End previous result parsing
+
+	switch {
+	case conf.Burst < 0 || conf.Rate < 0:
+		return nil, fmt.Errorf("rate and burst must be a positive integer")
+	case conf.Burst == 0 && conf.Rate != 0:
+		return nil, fmt.Errorf("if rate is set, burst must also be set")
+	case conf.Rate == 0 && conf.Burst != 0:
+		return nil, fmt.Errorf("if burst is set, rate must also be set")
+	}
+
+	return &conf, nil
+
+}
+
+func getIfbDeviceName(networkName string, containerId string) (string, error) {
+	hash := sha1.New()
+	_, err := hash.Write([]byte(networkName + containerId))
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%x", hash.Sum(nil))[:4], nil
+}
+
+func getMTU(deviceName string) (int, error) {
+	link, err := netlink.LinkByName(deviceName)
+	if err != nil {
+		return -1, err
+	}
+
+	return link.Attrs().MTU, nil
+}
+
+func getHostInterface(interfaces []*current.Interface) (*current.Interface, error) {
+	for _, prevIface := range interfaces {
+		if prevIface.Sandbox != "" {
+			continue
+		}
+
+		return prevIface, nil
+	}
+
+	return nil, errors.New("no host interface found")
+}
+
+// cmdAdd is called for ADD requests
+func cmdAdd(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	//no traffic shaping was requested, so just no-op and quit
+	if conf.Rate == 0 && conf.Burst == 0 {
+		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	hostInterface, err := getHostInterface(conf.PrevResult.Interfaces)
+	if err != nil {
+		return err
+	}
+
+	err = CreateIngressQdisc(conf.Rate, conf.Burst, hostInterface.Name)
+	if err != nil {
+		return err
+	}
+
+	mtu, err := getMTU(hostInterface.Name)
+	if err != nil {
+		return err
+	}
+
+	ifbDeviceName, err := getIfbDeviceName(conf.Name, args.ContainerID)
+	if err != nil {
+		return err
+	}
+
+	err = CreateIfb(ifbDeviceName, mtu)
+	if err != nil {
+		return err
+	}
+
+	ifbDevice, err := netlink.LinkByName(ifbDeviceName)
+	if err != nil {
+		return err
+	}
+
+	conf.PrevResult.Interfaces = append(conf.PrevResult.Interfaces, &current.Interface{
+		Name:    ifbDeviceName,
+		Sandbox: hostInterface.Sandbox,
```

Comment:
Sandbox should not be set, since the IFB device isn't in the sandbox / container netns.